feat: 거래내역 목록 조회

chinsanchung opened the pull request:
## 작업 분류

- [ ] 버그 수정
- [x] 신규 기능 추가
- [ ] 리팩터링
- [ ] 테스트

## 작업 개요
- 거래내역 목록 조회 작성

## 상세 내용
- UseGuards 와 연동하여 로그인 여부를 검증합니다.
- 요청에 전달할 쿼리는 account_id, page, trans_type, startDate, endDate 입니다. trans_type 은 "in", "out"이 있고, startDate 와 endDate 는 "YYYY-MM-DD" 형식으로 작성합니다.
    - startDate, endDate 을 쿼리에 작성하지 않았을 경우 3개월 전부터 오늘까지를 기본값으로 설정합니다.
    - trans_type 을 쿼리에 작성하지 않을 경우 입금과 출금 모두 출력하도록 작성했습니.
- 계좌 테이블과 연동하여, 계좌의 소유주인지 여부를 확인합니다. AccountRepository 를 서비스에 불러와 조회하는 방식으로 진행합니다.
- repository 에서는 계좌의 아이디, 거래일시와 입출금 여부를 필터링하고, 페이지네이션으로 1페이지에 5 항목을 출력하도록 설정했습니다.
- 거래일시("createdAt"), 거래금액("amount"), 잔액("balance"), 거래종류("trans_type"), 적요("comments")를 응답에 포함합니다.

## 집중 요망!
- moment-timezone 설치: 거래일시를 한국 시간으로 정하고, 시간 계산과 문자열 출력에 활용하기 위해 설치했습니다.

resolves: #14, #17

BangleCoding committed:
feat: 거래내역 전체 조회 진행중
- 거래내역 조건문 작성중, interface 작성, pagination 기능 완성
resolves #14

BangleCoding committed:
feat: Authguard 추가, getAllTransactions 기능 완성

chinsanchung committed:
feat: 거래내역 조회 목록 진행 - 날짜로 필터링, 계좌 테이블과 조인 수정

- 날짜 필터링 기능 추가
- moment-timezone을 설치해서 시작일과 마지막일의 기본값을 설정("3개월 전 ~ 오늘"으로 설정)
- transaction 엔티티에 eager 설정을 true 로 수정해서 거래내역 목록에 계좌의 아이디가 나오도록 설정

resolves: #14, #17

chinsanchung committed:
chore: sqlite 관련 불필요한 파일 정리

chinsanchung committed:
feat: 거래내역 조회 기능 및 moment-timezone 설치

- 유저 가드 적용
- 계좌의 소유주인지 여부를 검증
- 거래시간(시작 ~ 끝) 조건문 작성. 기본값은 3개월 전 ~ 현재 입니다.
- 서울 기준 설정, 거래시간 계산과 문자열 변환을 위해 moment-timezone 설치
- 페이지네이션
- 거래일시, 거래금액, 잔액, 거래종류, 적요만을 응답으로 보냅니다.

resolves: #14

chinsanchung committed:
fix: 리스트 메소드의 쿼리 인터페이스 수정

- trans_type 을 string 에서 'in' | 'out' 으로 수정

resolves: #14

chinsanchung committed:
fix: AccountsRepository 를 불러오는 경로 수정

- 경로에서 "src"를 ".."으로 수정했습니다.

resolves: #14

chinsanchung committed:
fix: Account 엔티티를 불러오는 경로 수정

- 경로에서 "src"를 "../.."으로 수정했습니다.

resolves: #14

yhshin0 commented:
입출금 전체 조회는 불가능한가요?

yhshin0 commented on src/users/users.controller.ts:
```diff
@@ -9,7 +9,7 @@ export class UsersController {
 
   @Post()
-  async createUser(@Body() createUserDto: CreateUserDto): Promise<User> {
+  async createUser(@Body() createUserDto): Promise<User> {
```
여기 createUserDto의 타입은 왜 제거하신건가요?

chinsanchung replied:
처음에 목록을 조회하기 전, 회원가입을 진행할 때 에러를 확인하려고 지웠는데 다시 복구하겠습니다.

chinsanchung replied:
그리고 위에 입출금 조회를 언급하셨는데, 현재 입금 또는 출금만 작성했습니다. 입출금 둘 다 출력하는 것도 작성하겠습니다.

yhshin0 commented on src/users/users.service.ts:
```diff
@@ -35,6 +35,7 @@ export class UsersService {
     const user = this.usersRepository.create(createUserDto);
     try {
       const result = await this.usersRepository.save(user);
+      console.log('we can signup');
```
주석은 지우는 것이 나을 것 같습니다

chinsanchung replied:
이 부분도 회원가입 에러 확인용으로 작성했는데 삭제하겠습니다.

yhshin0 commented on src/transaction/transaction.service.ts:
```diff
@@ -1,4 +1,43 @@
+
+  async getAllTransactions(query: PagingOptions): Promise<Transaction[]> {
+    // * 계좌의 소유주인지 여부를 확인합니다.
+    const account = await this.accountRepository.findOne(+query.account_id, {
+      join: {
+        alias: 'account',
+        leftJoinAndSelect: { user: 'account.user' },
+      },
+    });
+    if (!account) {
+      throw new BadRequestException(
+        '거래 내역에 등록한 계좌가 존재하지 않습니다.',
+      );
+    }
+    if (account.user.user_id !== query.user.user_id) {
+      throw new NotAcceptableException(
+        '오직 계좌의 소유주만 해당 계좌의 거래 내역을 조회하실 수 있습니다.',
+      );
+    }
```
account repository에 정의된 findOneByAccountNumber(acc_num: string, user: User) 메소드를 사용하는게 나아 보입니다

chinsanchung replied:
findOneByAccountNumber 메소드가 없는 브랜치에서 작성했는데, 다시 받아온 후에 진행할게요.

chinsanchung replied:
findOneByAccountNumber 메소드를 확인했는데, 메소드의 에러 처리가 제가 생각한 방식과 다른 점이 있어서 기존의 방식으로 사용할까 합니다.

yhshin0 commented on src/transaction/transaction.interface.ts:
```diff
@@ -0,0 +1,15 @@
+import { User } from '../users/entities/user.entity';
+
+export interface ListQueryOptions {
+  account_id?: string;
+  page: string;
```
account_id는 필수 아닌가요?
그리고 page가 option으로 보입니다.

chinsanchung replied:
계좌 아이디의 물음표는 계좌를 생성하기 전 조회를 할 당시에 했던 설정인데 수정하겠습니다.
page 가 option 이라는 것은 이해를 하지 못했습니다.

yhshin0 replied:
아닙니다 잘못 봤네요

yhshin0 reviewed (commented):
입금/출금 조회는 되는 것 같은데 입출금 조회도 가능한가요?

chinsanchung committed:
fix: PR 피드백 반영 및 service, repository 에서의 매개변수 인터페이스 이름 수정

- ListQueryOptions 인터페이스의 계좌 아이디의 물음표 삭제
- 유저 컨트롤러의 createUser에서 매개변수의 dto 를 다시 되살림
- 유저 서비스의 createUser에서 콘솔 로그 삭제
- PagingOptions => ListWithPageAndUserOptions 으로 인터페이스 이름 수정

resolves: #14

chinsanchung committed:
Merge branch 'develop' into feature/transaction-list

chinsanchung committed:
fix: 입금 출금 필터링 수정

- 입금, 출금뿐만 아니라 입출금 모두 출력하도록 수정했습니다.
- 요청할 때 쿼리에 trans_type 을 제거하면 입출금 모두를 출력합니다.

resolves: #14

chinsanchung committed:
Merge branch 'develop' into feature/transaction-list
